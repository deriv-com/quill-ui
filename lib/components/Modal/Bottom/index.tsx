import ReactDOM from "react-dom";
import { useState, useEffect, useRef } from "react";
import clsx from "clsx";
import "./modal-bottom.scss";
import { useSwipeable } from "react-swipeable";
import { ModalImage } from "./modal-image";
import { ModalTitle } from "./modal-title";
import { ModalBody } from "./modal-body";
import { Button } from "../../Button";

interface ModalBottomProps {
    isOpened?: boolean;
    className?: string;
    showHandleBar?: boolean;
    showSecondaryButton?: boolean;
    shouldCloseOnPrimaryButtonClick?: boolean;
    toggleModal: (isOpened: boolean) => void;
    portalId?: string;
    primaryButtonLabel?: React.ReactNode;
    primaryButtonCallback?: () => void;
    secondaryButtonLabel?: React.ReactNode;
}

const swipeConfig = {
    delta: 10,
    trackTouch: true,
    trackMouse: true,
    preventScrollOnSwipe: true,
};

const MAX_HEIGHT = 85;

export const ModalBottom = ({
    isOpened = false,
    className,
    children,
    showHandleBar = true,
    showSecondaryButton = true,
    shouldCloseOnPrimaryButtonClick = false,
    toggleModal,
    portalId,
    primaryButtonLabel,
    primaryButtonCallback,
    secondaryButtonLabel,
}: React.PropsWithChildren<ModalBottomProps>) => {
    const [isVisible, setIsVisible] = useState(isOpened);
    const [isExpanded, setIsExpanded] = useState(false);
    const [isSwiping, setIsSwiping] = useState(false);

    const animationTimerRef = useRef<ReturnType<typeof setTimeout>>();
    const swipingTimerRef = useRef<ReturnType<typeof setTimeout>>();
    const scrollableContainerRef = useRef<HTMLDivElement>(null);

    const modalRoot =
        (portalId && document.getElementById(portalId)) ||
        document.getElementById("modal-root") ||
        document.body;

    const shouldExpand =
        scrollableContainerRef?.current?.offsetHeight && !isExpanded
            ? Math.round(
                  (scrollableContainerRef?.current?.offsetHeight /
                      window.innerHeight) *
                      100,
              ) < MAX_HEIGHT
            : true;

    useEffect(() => {
        setIsVisible(isOpened);
    }, [isOpened]);

    useEffect(() => {
        return () => {
            clearTimeout(animationTimerRef.current);
            clearTimeout(swipingTimerRef.current);
        };
    }, []);

    const toggleHandler = () => {
        setIsVisible(!isVisible);
        animationTimerRef.current = setTimeout(() => {
            setIsExpanded(false);
            toggleModal(!isOpened);
        }, 300);
    };

    const primaryButtonFunctionHandler = () => {
        primaryButtonCallback?.();
        if (shouldCloseOnPrimaryButtonClick) toggleHandler();
    };

    const swipeHandlers = useSwipeable({
        onSwipedUp: () => (shouldExpand ? setIsExpanded(true) : null),
        onSwipedDown: () =>
            shouldExpand ? setIsExpanded(false) : toggleHandler(),
        onSwipeStart: () => setIsSwiping(true),
        onSwiped: () =>
            (swipingTimerRef.current = setTimeout(
                () => setIsSwiping(false),
                300,
            )),
        ...swipeConfig,
    });
    if (!isOpened) return null;

    return ReactDOM.createPortal(
        <div
            className="quill-modal-bottom__overlay"
            onClick={isSwiping ? undefined : toggleHandler}
        >
            <div
                className={clsx(
                    "quill-modal-bottom__container",
                    {
                        "quill-modal-bottom__container--visible": isVisible,
                        "quill-modal-bottom__container--expanded": isExpanded,
                    },
                    className,
                )}
                onClick={(e) => e.stopPropagation()}
                ref={scrollableContainerRef}
            >
                {showHandleBar && (
                    <div
                        className="quill-modal-bottom__handle-bar"
                        {...swipeHandlers}
                    />
                )}
                <div className="quill-modal-bottom__content-wrapper">
                    {children}
                </div>
                <div className="quill-modal-bottom__button-wrapper">
                    <Button
                        color="black"
                        fullWidth
                        size="lg"
                        label={primaryButtonLabel}
                        onClick={primaryButtonFunctionHandler}
                    />
                    {showSecondaryButton && (
                        <Button
                            color="black"
                            fullWidth
                            size="lg"
                            label={secondaryButtonLabel}
                            variant="secondary"
                            className="quill-modal-bottom__button"
                            onClick={toggleHandler}
                        />
                    )}
                </div>
            </div>
        </div>,
        modalRoot,
    );
};

ModalBottom.Title = ModalTitle;
ModalBottom.Body = ModalBody;
ModalBottom.Image = ModalImage;
