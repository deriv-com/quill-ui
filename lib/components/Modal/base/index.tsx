import ReactDOM from "react-dom";
import { useState, useEffect, useRef, HTMLAttributes } from "react";
import clsx from "clsx";
import { Button, IconButton } from "@components/Button";
import { LabelPairedXmarkMdBoldIcon } from "@deriv/quill-icons/LabelPaired";
import { ModalHeader } from "./modal-header";
import { ModalBody } from "./modal-body";
import "../modal.scss";
import useBreakpoints from "@hooks/useBreakpoints";
import ModalActionSheet, { ModalActionSheetProps } from "./modal-action-sheet";

export interface ModalProps extends HTMLAttributes<HTMLDivElement> {
    isOpened?: boolean;
    className?: string;
    containerClassName?: string;
    showHandleBar?: boolean;
    showCrossIcon?: boolean;
    showSecondaryButton?: boolean;
    shouldCloseOnPrimaryButtonClick?: boolean;
    shouldCloseOnSecondaryButtonClick?: boolean;
    disableCloseOnOverlay?: boolean;
    toggleModal?: (isOpened: boolean) => void;
    portalId?: string;
    showPrimaryButton?: boolean;
    primaryButtonLabel?: React.ReactNode;
    isPrimaryButtonDisabled?: boolean;
    isSecondaryButtonDisabled?: boolean;
    primaryButtonCallback?: () => void;
    secondaryButtonCallback?: () => void;
    secondaryButtonLabel?: React.ReactNode;
    isMobile?: boolean;
    hasFooter?: boolean;
    actionSheetProps?: ModalActionSheetProps;
}

export const Modal = ({
    isOpened = false,
    className,
    children,
    showHandleBar,
    showCrossIcon,
    showPrimaryButton = true,
    showSecondaryButton = false,
    shouldCloseOnPrimaryButtonClick = false,
    shouldCloseOnSecondaryButtonClick = false,
    toggleModal,
    disableCloseOnOverlay = false,
    hasFooter = true,
    portalId,
    primaryButtonLabel,
    isPrimaryButtonDisabled,
    isSecondaryButtonDisabled,
    primaryButtonCallback,
    secondaryButtonCallback,
    secondaryButtonLabel,
    actionSheetProps,
    ...rest
}: React.PropsWithChildren<ModalProps>) => {
    const [isVisible, setIsVisible] = useState(isOpened);
    const [isExpanded, setIsExpanded] = useState(false);
    // const [isSwiping, setIsSwiping] = useState(disableCloseOnOverlay);

    // const animationTimerRef = useRef<ReturnType<typeof setTimeout>>();
    // const swipingTimerRef = useRef<ReturnType<typeof setTimeout>>();
    const scrollableContainerRef = useRef<HTMLDivElement>(null);
    const actionSheetRef = useRef<HTMLDivElement>(null);

    const { isMobile } = useBreakpoints();

    const modalRoot =
        (portalId && document.getElementById(portalId)) ||
        document.getElementById("modal-root") ||
        document.body;

    useEffect(() => {
        setIsVisible(isOpened);
    }, [isOpened]);

    // useEffect(() => {
    //     return () => {
    //         clearTimeout(animationTimerRef.current);
    //         clearTimeout(swipingTimerRef.current);
    //     };
    // }, []);

    const toggleHandler = () => {
        setIsVisible((prev) => !prev);

        // animationTimerRef.current = setTimeout(() => {
        //     setIsExpanded(false);
        //     toggleModal && toggleModal(!isOpened);
        // }, 300);
    };

    const primaryButtonFunctionHandler = () => {
        primaryButtonCallback?.();
        if (shouldCloseOnPrimaryButtonClick) toggleHandler();
    };
    const secondaryButtonFunctionHandler = () => {
        secondaryButtonCallback?.();
        if (shouldCloseOnSecondaryButtonClick) toggleHandler();
    };

    const {
        root = {},
        portal = {},
        ...restActionSheetProps
    } = actionSheetProps || {};

    const handleClose = () => {
        if (disableCloseOnOverlay) return;
        if (toggleModal && isVisible) return toggleModal(false);
        setIsVisible(false);
    };

    const updatedRoot = {
        ...root,
        isOpen: isVisible,
        onClose: handleClose,
    };

    console.log(isVisible);

    const updatedPortal = {
        ...portal,
        showHandleBar,
    };

    return !isMobile && isVisible ? (
        ReactDOM.createPortal(
            <div
                {...rest}
                className="quill-modal__background"
                data-testid="dt_overlay"
                // onClick={isSwiping ? undefined : toggleHandler}
                onClick={handleClose}
            >
                <div
                    className={clsx(
                        `quill-modal__container${isMobile ? "--mobile" : "--desktop"}`,
                        {
                            "quill-modal__container--visible": isVisible,
                            "quill-modal__container--expanded": isExpanded,
                        },

                        className,
                    )}
                    onClick={(e) => e.stopPropagation()}
                    ref={scrollableContainerRef}
                >
                    {showCrossIcon && (
                        <IconButton
                            variant="tertiary"
                            color="black-white"
                            size="lg"
                            className="quill-modal__close-icon"
                            onClick={toggleHandler}
                            icon={<LabelPairedXmarkMdBoldIcon />}
                        />
                    )}

                    <div className="quill-modal__content-wrapper">
                        {children}
                    </div>
                    {hasFooter && (
                        <div className="quill-modal__button-wrapper">
                            {showPrimaryButton && (
                                <Button
                                    color="black-white"
                                    fullWidth
                                    size="lg"
                                    label={primaryButtonLabel}
                                    disabled={isPrimaryButtonDisabled}
                                    onClick={primaryButtonFunctionHandler}
                                />
                            )}
                            {showSecondaryButton && (
                                <Button
                                    color="black-white"
                                    fullWidth
                                    size="lg"
                                    label={secondaryButtonLabel}
                                    disabled={isSecondaryButtonDisabled}
                                    variant="secondary"
                                    className="quill-modal__button"
                                    onClick={secondaryButtonFunctionHandler}
                                />
                            )}
                        </div>
                    )}
                </div>
            </div>,
            modalRoot,
        )
    ) : (
        <ModalActionSheet
            root={updatedRoot}
            portal={updatedPortal}
            ref={actionSheetRef}
            {...restActionSheetProps}
        >
            {children}
        </ModalActionSheet>
    );
};
Modal.Header = ModalHeader;
Modal.Body = ModalBody;
