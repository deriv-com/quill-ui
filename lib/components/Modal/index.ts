export * from "./bottom";
export * from "./overlay";
