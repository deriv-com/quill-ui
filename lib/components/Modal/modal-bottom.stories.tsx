import type { Meta, StoryObj } from "@storybook/react";
import { fn } from "@storybook/test";
import { useEffect, useState } from "react";
import {
    StandaloneCircleSterlingRegularIcon,
    StandaloneTrashRegularIcon,
} from "@deriv/quill-icons";
import { Modal } from "./index";
import { Button } from "@components/Button";

interface Template extends React.ComponentProps<typeof Modal> {
    image?: React.ReactNode;
    src?: string;
    style?: React.CSSProperties;
    textContent?: React.ReactNode;
}

const openModalButtonLabel = "Open Modal";
const primaryButtonLabel = "Primary Button Label";
const secondaryButtonLabel = "Secondary Button Label";
const shortTextContent = "This is some amazing placeholder.";
const placeHolderText =
    "Lorem ipsum dolor sit amet consectetur. Venenatis malesuada nibh sed ornare rnare id suspendisse sed.";
const mediumTextContent = placeHolderText.padStart(200, placeHolderText);
const longTextContent = placeHolderText.padStart(600, placeHolderText);
const titlePlaceHolderText = "Title";
const imageSRC =
    "https://live.staticflickr.com/603/21947667154_e63cc9252b_b.jpg";
const ImageComponent = <img src={imageSRC} alt="Apples" />;

const preloadImage = (imageSRC: string) => {
    const img = new Image();
    img.src = imageSRC;
};

preloadImage(imageSRC);

const meta = {
    title: "Components/Modal/Bottom",
    component: Modal,
    tags: ["autodocs"],
    parameters: {
        layout: "centered",
        viewport: {
            defaultViewport: "mobile1",
        },
    },
    args: {
        children: <div>{shortTextContent}</div>,
        isOpened: false,
        disableCloseOnOverlay: false,
        showHandleBar: true,
        isMobile: true,
        showSecondaryButton: true,
        shouldCloseOnPrimaryButtonClick: false,
        toggleModal: fn(),
        primaryButtonLabel: primaryButtonLabel,
        primaryButtonCallback: fn(),
        secondaryButtonLabel: secondaryButtonLabel,
    },
    argTypes: {
        children: {
            table: { type: { summary: "ReactNode" } },
            description:
                "Modal's content. Can be wrapped with the `<ModalBottom.Header/>` and `<ModalBottom.Body/>` components in order to organize the content inside the modal. Each of them accepts className for customization and `<ModalBottom.Header/>` can also be passed scr and height properties.",
            control: { type: null },
        },

        isOpened: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            description: "Controls the visibility of the modal",
            control: { type: "boolean" },
        },
        className: {
            table: { type: { summary: "string | undefined" } },
            description: "ClassName for external tag of the component.",
            control: { type: "text" },
        },
        showHandleBar: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            description:
                "Controls the visibility of the handlebar. If handlebar is visible, the modal can be expanded/collapsed by swiping up/down handlebar. Default value is true.",
            control: { type: "boolean" },
        },
        showSecondaryButton: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            description: "Controls the visibility of the secondary button.",
            control: { type: "boolean" },
        },
        shouldCloseOnPrimaryButtonClick: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            description:
                "Flag for controlling modal behavior. If it's true, then the modal will be closed after user clicks on the primary button.",
            control: { type: "boolean" },
        },
        toggleModal: {
            table: { type: { summary: "(isOpened: boolean) => void" } },
            description:
                "Function for changing state of the visibility of the modal.",
            control: { type: null },
        },
        portalId: {
            table: { type: { summary: "string | undefined" } },
            description:
                "ID of the modal container. If it wasn't passed, the modal will try to find any container with 'modal-root' ID. If there is no container with 'modal-root' ID the modal will be insert inside document.body.",
            control: { type: "string" },
        },
        primaryButtonLabel: {
            table: { type: { summary: "ReactNode" } },
            description: "Label for the primary button.",
            control: { type: "string" },
        },
        primaryButtonCallback: {
            table: { type: { summary: "(() => void) | undefined" } },
            description:
                "Function which will be called on clicking on primary button.",
            control: { type: null },
        },
        secondaryButtonLabel: {
            table: { type: { summary: "ReactNode | undefined" } },
            description: "Label for the secondary button.",
            control: { type: "string" },
        },
    },
} satisfies Meta<typeof Modal>;

export default meta;

type Story = StoryObj<typeof meta>;

const Template: React.FC<Template> = ({
    image,
    src,
    style,
    textContent = shortTextContent,
    ...args
}: Template) => {
    const [isOpen, setIsOpen] = useState(args.isOpened);

    useEffect(() => {
        setIsOpen(args.isOpened);
    }, [args.isOpened]);

    return (
        <>
            <Button
                size="lg"
                label={openModalButtonLabel}
                onClick={() => setIsOpen(true)}
            />
            <Modal {...args} isOpened={isOpen} toggleModal={setIsOpen}>
                <Modal.Header
                    title={titlePlaceHolderText}
                    image={image}
                    src={src}
                    style={style}
                />
                <Modal.Body>{textContent}</Modal.Body>
            </Modal>
        </>
    );
};

export const DefaultModalBottom = Template.bind(this) as Story;
DefaultModalBottom.args = { ...meta.args };

export const ModalBottomWithoutHandleBar = Template.bind(this) as Story;
ModalBottomWithoutHandleBar.args = {
    ...meta.args,
    showHandleBar: false,
};
export const DisableCloseOnOverlay = Template.bind(this) as Story;
DisableCloseOnOverlay.args = {
    ...meta.args,
    disableCloseOnOverlay: true,
};

export const ModalBottomWithoutSecondaryButton = Template.bind(this) as Story;
ModalBottomWithoutSecondaryButton.args = {
    ...meta.args,
    showSecondaryButton: false,
};

export const ClosingModalBottomOnPrimaryButtonClick = Template.bind(
    this,
) as Story;
ClosingModalBottomOnPrimaryButtonClick.args = {
    ...meta.args,
    shouldCloseOnPrimaryButtonClick: true,
    textContent: mediumTextContent,
} as Story["args"];

export const ModalBottomExpandedByDefault = Template.bind(this) as Story;
ModalBottomExpandedByDefault.args = {
    ...meta.args,
    textContent: longTextContent,
} as Story["args"];

export const ModalBottomWithImage = Template.bind(this) as Story;
ModalBottomWithImage.args = {
    ...meta.args,
    image: ImageComponent,
} as Story["args"];

export const ModalBottomWithImageAndLongContent = Template.bind(this) as Story;
ModalBottomWithImageAndLongContent.args = {
    ...meta.args,
    image: ImageComponent,
    textContent: longTextContent,
} as Story["args"];

export const ModalBottomWithImageSrc = Template.bind(this) as Story;
ModalBottomWithImageSrc.args = {
    ...meta.args,
    src: imageSRC,
    textContent: mediumTextContent,
} as Story["args"];

export const ModalBottomWithIcon = Template.bind(this) as Story;
ModalBottomWithIcon.args = {
    ...meta.args,
    image: (
        <StandaloneTrashRegularIcon
            fill="var(--core-color-solid-red-900)"
            iconSize="2xl"
        />
    ),
    style: {
        backgroundColor: "var(--core-color-solid-red-100)",
    },
} as Story["args"];

export const ModalBottomWithIconAndLongContent = Template.bind(this) as Story;
ModalBottomWithIconAndLongContent.args = {
    ...meta.args,
    image: (
        <StandaloneCircleSterlingRegularIcon
            fill="var(--core-color-solid-green-900)"
            iconSize="2xl"
        />
    ),
    style: {
        backgroundColor: "var(--core-color-solid-green-100)",
    },
    textContent: longTextContent,
} as Story["args"];
