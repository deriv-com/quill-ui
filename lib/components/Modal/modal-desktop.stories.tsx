import type { Meta, StoryObj } from "@storybook/react";
import { ModalExample } from "./mocks/example";
import { StandaloneTrashRegularIcon } from "@deriv/quill-icons/Standalone";
const placeHolderText =
    "Lorem ipsum dolor sit amet consectetur. Venenatis malesuada nibh sed ornare rnare id suspendisse sed.";
const mediumTextContent = placeHolderText.padStart(200, placeHolderText);
const longTextContent = placeHolderText.padStart(2000, placeHolderText);
const shortTextContent = "This is a description content.";
const imageSRC =
    "https://live.staticflickr.com/603/21947667154_e63cc9252b_b.jpg";
const ImageComponent = <img src={imageSRC} alt="Apples" />;

const preloadImage = (imageSRC: string) => {
    const img = new Image();
    img.src = imageSRC;
};

preloadImage(imageSRC);

const meta = {
    title: "Components/Modal/Overlay",
    component: ModalExample,
    tags: ["autodocs"],
    parameters: {
        layout: "centered",
    },
    args: {
        showHandleBar: true,
        showSecondaryButton: true,
        disableCloseOnOverlay: false,
    },
    argTypes: {
        children: {
            table: { type: { summary: "ReactNode" } },
            description:
                "Modal's content. Can be wrapped with the `<ModalOverlay.Header/>` and `<ModalOverlay.Body/>` components in order to organize the content inside the modal. Each of them accepts className for customization and `<ModalOverlay.Header/>` can also be passed scr and height properties.",
            control: { type: null },
        },
        isOpened: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            description: "Controls the visibility of the modal",
            control: { type: "boolean" },
        },

        className: {
            table: { type: { summary: "string | undefined" } },
            description: "ClassName for external tag of the component.",
            control: { type: "text" },
        },
        isMobile: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            description:
                "Controls the visibility of the handlebar. If handlebar is visible, the modal can be expanded/collapsed by swiping up/down handlebar. Default value is true.",
            control: { type: "boolean" },
        },
        showSecondaryButton: {
            table: { type: { summary: "boolean | undefined" } },
            options: ["true", "false"],
            description: "Controls the visibility of the secondary button.",
            control: { type: "boolean" },
        },

        toggleModal: {
            table: { type: { summary: "(isOpened: boolean) => void" } },
            description:
                "Function for changing state of the visibility of the modal.",
            control: { type: null },
        },
        portalId: {
            table: { type: { summary: "string | undefined" } },
            description:
                "ID of the modal container. If it wasn't passed, the modal will try to find any container with 'modal-root' ID. If there is no container with 'modal-root' ID the modal will be insert inside document.body.",
            control: { type: "string" },
        },
        primaryButtonLabel: {
            table: { type: { summary: "ReactNode" } },
            description: "Label for the primary button.",
            control: { type: "string" },
        },
        primaryButtonCallback: {
            table: { type: { summary: "(() => void) | undefined" } },
            description:
                "Function which will be called on clicking on primary button.",
            control: { type: null },
        },
        secondaryButtonLabel: {
            table: { type: { summary: "ReactNode | undefined" } },
            description: "Label for the secondary button.",
            control: { type: "string" },
        },
    },
} satisfies Meta<typeof ModalExample>;

export default meta;

type Story = StoryObj<typeof ModalExample>;

export const Default: Story = {
    args: {
        textContent: shortTextContent,
        showCrossIcon: true,
    },
};
export const ModalOverlayWithoutCrossIcon: Story = {
    args: {
        textContent: shortTextContent,
    },
};
export const ModalOverlayWithoutSecondaryButton: Story = {
    args: {
        showSecondaryButton: false,
        textContent: shortTextContent,
        showCrossIcon: true,
    },
};
export const ClosingModalOverlayOnPrimaryButtonClick: Story = {
    args: {
        shouldCloseOnPrimaryButtonClick: true,
        textContent: mediumTextContent,
        showCrossIcon: true,
    },
};
export const ModalOverlayWithLongContent: Story = {
    args: {
        textContent: longTextContent,
        showCrossIcon: true,
    },
};

export const ModalOverlayWithImage: Story = {
    args: {
        image: ImageComponent,
        showCrossIcon: true,
    },
};
export const ModalOverlayWithImageAndLongContent: Story = {
    args: {
        image: ImageComponent,
        textContent: longTextContent,
        showCrossIcon: true,
    },
};
export const ModalOverlayWithImageSrc: Story = {
    args: {
        src: imageSRC,
        textContent: mediumTextContent,
        showCrossIcon: true,
    },
};
export const ModalOverlayWithIcon: Story = {
    args: {
        showCrossIcon: true,
        image: (
            <StandaloneTrashRegularIcon
                fill="var(--core-color-solid-red-900)"
                iconSize="2xl"
            />
        ),
        style: {
            backgroundColor: "var(--core-color-solid-red-100)",
        },
    },
};
export const ModalOverlayWithIconAndLongContent: Story = {
    args: {
        showCrossIcon: true,
        image: (
            <StandaloneTrashRegularIcon
                fill="var(--core-color-solid-red-900)"
                iconSize="2xl"
            />
        ),
        style: {
            backgroundColor: "var(--core-color-solid-red-100)",
        },
        textContent: longTextContent,
    },
};
