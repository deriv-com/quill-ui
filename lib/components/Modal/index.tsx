import React, { ReactNode } from "react";
import { ModalHeader } from "./base/modal-header";
import { ModalBody } from "./base/modal-body";
import { ModalBase } from "./base/base";

export interface ModalType {
    Header: typeof ModalHeader;
    Body: typeof ModalBody;
    Frame: typeof ModalBase;
}

export const Modal: ModalType = ({ children }: { children: ReactNode }) => {
    return <>{children}</>;
};

Modal.Frame = ModalBase;
Modal.Header = ModalHeader;
Modal.Body = ModalBody;

export default Modal;
