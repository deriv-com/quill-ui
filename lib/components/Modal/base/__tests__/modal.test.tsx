import React from "react";
import ReactDOM from "react-dom";
import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { StandaloneTrashRegularIcon } from "@deriv/quill-icons";
import Modal from "@components/Modal/base";

describe("Modal.Base", () => {
    const primaryButtonLabel = "Primary Button Label";
    const secondaryButtonLabel = "Secondary Button Label";
    const shortTextContent = "This is some amazing placeholder.";
    const title = "Title";
    const imageSrc =
        "https://live.staticflickr.com/603/21947667154_e63cc9252b_b.jpg";

    const children = (
        <>
            <Modal.Header title={title} />
            <Modal.Body>{shortTextContent}</Modal.Body>
        </>
    );

    const mockProps = {
        isOpened: true,
        toggleModal: jest.fn(),
        primaryButtonLabel,
    };

    beforeAll(() => {
        (ReactDOM.createPortal as jest.Mock) = jest.fn(
            (component) => component,
        );
    });

    afterAll(() => {
        (ReactDOM.createPortal as jest.Mock).mockClear();
    });

    it("should render with default values if optional ones were not passed", () => {
        const { container } = render(
            <Modal.Base {...mockProps}>{children}</Modal.Base>,
        );

        expect(container).toMatchSnapshot();
    });

    it("should call toggleModal if user clicks on the overlay", async () => {
        render(<Modal.Base {...mockProps}>{children}</Modal.Base>);

        userEvent.click(screen.getByTestId("dt_overlay"));
        await waitFor(() => {
            expect(mockProps.toggleModal).toHaveBeenCalledTimes(1);
        });
    });

    it("should render secondary button if showSecondaryButton === true and the label was passed", () => {
        const { container } = render(
            <Modal.Base
                {...mockProps}
                showSecondaryButton
                secondaryButtonLabel={secondaryButtonLabel}
            >
                {children}
            </Modal.Base>,
        );

        expect(screen.getByText(secondaryButtonLabel)).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should call toggleModal if user clicks on secondary button", async () => {
        render(
            <Modal.Base
                {...mockProps}
                showSecondaryButton
                secondaryButtonLabel={secondaryButtonLabel}
            >
                {children}
            </Modal.Base>,
        );

        userEvent.click(screen.getByText(secondaryButtonLabel));
        await waitFor(() => {
            expect(mockProps.toggleModal).toHaveBeenCalledTimes(1);
        });
    });

    it("should call toggleModal if user clicks on primary button and shouldCloseOnPrimaryButtonClick == true", async () => {
        render(
            <Modal.Base {...mockProps} shouldCloseOnPrimaryButtonClick>
                {children}
            </Modal.Base>,
        );

        userEvent.click(screen.getByText(primaryButtonLabel));
        await waitFor(() => {
            expect(mockProps.toggleModal).toHaveBeenCalledTimes(1);
        });
    });

    it("should call primaryButtonCallback if user clicks on primary button and primaryButtonCallback was passed", async () => {
        const primaryButtonCallback = jest.fn();
        render(
            <Modal.Base
                {...mockProps}
                primaryButtonCallback={primaryButtonCallback}
            >
                {children}
            </Modal.Base>,
        );

        userEvent.click(screen.getByText(primaryButtonLabel));
        await waitFor(() => {
            expect(primaryButtonCallback).toHaveBeenCalledTimes(1);
        });
    });

    it("should render container with image as a background if src was passed", () => {
        const { container } = render(
            <Modal.Base {...mockProps}>
                <Modal.Header src={imageSrc} title={title} />
            </Modal.Base>,
        );

        expect(screen.getByTestId("dt_modal_image")).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it("should render SVG if it was passed", () => {
        const { container } = render(
            <Modal.Base {...mockProps}>
                <Modal.Header
                    title={title}
                    image={<StandaloneTrashRegularIcon />}
                />
            </Modal.Base>,
        );

        expect(container).toMatchSnapshot();
    });
});
