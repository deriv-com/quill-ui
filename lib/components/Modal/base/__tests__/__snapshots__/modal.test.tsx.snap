// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal should render SVG if it was passed 1`] = `
<div>
  <div
    class="quill-modal__background"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal__container--desktop quill-modal__container--visible"
    >
      <div
        class="quill-modal__content-wrapper"
      >
        <div
          class="quill-modal__content-image"
        >
          <svg
            height="32"
            role="img"
            viewBox="0 0 32 32"
            width="32"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M14.164 7.75a.61.61 0 0 0-.508.313L13.031 9h5.899l-.625-.937a.61.61 0 0 0-.508-.313zM20.375 9h3.75c.313 0 .625.313.625.625a.64.64 0 0 1-.625.625h-.742l-.977 13.945c-.117 1.328-1.172 2.305-2.5 2.305h-7.851a2.51 2.51 0 0 1-2.5-2.305L8.578 10.25h-.703a.617.617 0 0 1-.625-.625c0-.312.273-.625.625-.625h3.711l1.016-1.602a1.84 1.84 0 0 1 1.562-.898h3.633c.625 0 1.25.352 1.601.898zm1.758 1.25H9.828l.977 13.867c.039.625.586 1.133 1.25 1.133h7.851c.664 0 1.211-.508 1.25-1.133z"
            />
          </svg>
        </div>
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal__content-title-has-image"
        >
          Title
        </h4>
      </div>
      <div
        class="quill-modal__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="quill-button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal should render container with image as a background if src was passed 1`] = `
<div>
  <div
    class="quill-modal__background"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal__container--desktop quill-modal__container--visible"
    >
      <div
        class="quill-modal__content-wrapper"
      >
        <div
          class="quill-modal__content-image quill-modal__content-image-size"
          data-testid="dt_modal_image"
        />
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal__content-title-has-image"
        >
          Title
        </h4>
      </div>
      <div
        class="quill-modal__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="quill-button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal should render secondary button if showSecondaryButton === true and the label was passed 1`] = `
<div>
  <div
    class="quill-modal__background"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal__container--desktop quill-modal__container--visible"
    >
      <div
        class="quill-modal__content-wrapper"
      >
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal__content-title"
        >
          Title
        </h4>
        <div
          class="quill-typography__body-text__size--md__weight--regular__decoration--default quill-typography__color--default quill-modal__content-body"
        >
          This is some amazing placeholder.
        </div>
      </div>
      <div
        class="quill-modal__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="quill-button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
        <button
          class="quill-button quill-button__size--lg quill__color--secondary-black quill-modal__button quill-button__full-width"
          data-state=""
        >
          <span
            class="quill-button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Secondary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`Modal should render with default values if optional ones were not passed 1`] = `
<div>
  <div
    class="quill-modal__background"
    data-testid="dt_overlay"
  >
    <div
      class="quill-modal__container--desktop quill-modal__container--visible"
    >
      <div
        class="quill-modal__content-wrapper"
      >
        <h4
          class="quill-typography__h4 quill-typography__color--prominent quill-modal__content-title"
        >
          Title
        </h4>
        <div
          class="quill-typography__body-text__size--md__weight--regular__decoration--default quill-typography__color--default quill-modal__content-body"
        >
          This is some amazing placeholder.
        </div>
      </div>
      <div
        class="quill-modal__button-wrapper"
      >
        <button
          class="quill-button quill-button__size--lg quill__color--primary-black quill-button__full-width"
          data-state=""
        >
          <span
            class="quill-button-label"
          >
            <p
              class="quill-typography__body-text__size--md__weight--bold__decoration--default black"
            >
              Primary Button Label
            </p>
          </span>
        </button>
      </div>
    </div>
  </div>
</div>
`;
